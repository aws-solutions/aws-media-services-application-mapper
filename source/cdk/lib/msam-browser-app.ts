/*! Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
       SPDX-License-Identifier: Apache-2.0 */

import {
    Aws,
    aws_cloudfront as cloudfront,
    aws_iam as iam,
    aws_lambda as lambda,
    aws_s3 as s3,
    CustomResource,
    CustomResourceProps,
    Duration,
    Fn,
    NestedStack,
    NestedStackProps,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as utils from './utils';

interface MsamBrowserAppProps extends NestedStackProps {
    /**
     * This is the IAM Role ARN for the Web app Lambda functions.
     */
    readonly WebIAMRole: iam.Role;
}

export class MsamBrowserApp extends NestedStack {
    /**
     * URL for the MSAM browser application via CloudFront
     */
    readonly MSAMBrowserURL: string;

    /**
     * MSAM browser application bucket
     */
    readonly MSAMBrowserAppBucket: string;

    constructor(scope: Construct, id: string, props: MsamBrowserAppProps) {
        super(
            scope,
            id,
            {
                ...props,
                description: 'Media Services Application Mapper Browser Application %%VERSION%%',
            }
        );

        /**
         * Resources
         */

        // Browser App Logging S3 Bucket
        const s3BucketbrowserAppLogs = new s3.Bucket(this, 'MSAMBrowserAppLoggingBucket', {
            accessControl: s3.BucketAccessControl.LOG_DELIVERY_WRITE,
            encryption: s3.BucketEncryption.S3_MANAGED,
            lifecycleRules: [
                {
                    abortIncompleteMultipartUploadAfter: Duration.days(5),
                    expiration: Duration.days(365),
                    id: 'Expire Objects After 1 Year',
                    noncurrentVersionExpiration: Duration.days(365),
                }
            ],
        });
        utils.setNagSuppressRules(s3BucketbrowserAppLogs, 
            {
                id: 'W35',
                id2: 'AwsSolutions-S1',
                reason: 'This is the logging bucket. No access logging needed.'
            },
            {
                id: 'W51',
                reason: 'This is the logging bucket.  No bucket policy needed.'
            },
            {
                id: 'AwsSolutions-S2',
                reason: 'Public access blocked by default'
            },
            {
                id: 'AwsSolutions-S10',
                reason: 'HTTPS requirement not supported at this time.',
            }
        );

        // Browser App S3 Bucket
        const s3BucketBrowserApp = new s3.Bucket(this, 'MSAMBrowserAppBucket', {
            cors: [
                {
                    allowedHeaders: ['*'],
                    allowedMethods: [s3.HttpMethods.GET],
                    allowedOrigins: ['*'],
                }
            ],
            encryption: s3.BucketEncryption.S3_MANAGED,
        });

        this.MSAMBrowserAppBucket = Fn.join('', [
            'https://s3.console.aws.amazon.com/s3/buckets/',
            s3BucketBrowserApp.bucketName,
            '/?region=',
            Aws.REGION,
        ]);

        utils.setNagSuppressRules(s3BucketBrowserApp,
            {
                id: 'AwsSolutions-S2',
                reason: 'Public access blocked by default'
            },
            {
                id: 'W51',
                reason: 'This is the logging bucket. No bucket policy needed.'
            },
            {
                id: 'AwsSolutions-S10',
                reason: 'HTTPS requirement not supported at this time.',
            }
        );

        (s3BucketBrowserApp.node.defaultChild as s3.CfnBucket).loggingConfiguration = {
            destinationBucketName: s3BucketbrowserAppLogs.bucketName,
            logFilePrefix: 'app-bucket/',
        };

        // CloudFront OriginAccessIdentity
        const cloudfrontOriginAccessIdentity = new cloudfront.OriginAccessIdentity(this, 'MSAMAppBucketOriginAccessIdentity', {
            comment: Fn.join('', [
                'Origin Access Identity for ',
                s3BucketBrowserApp.bucketName,
            ])
        });

        // Browser App S3 Bucket Policy
        const s3BucketAppPolicy = this.createS3BucketPolicy('MSAMBrowserAppBucketPolicy', {
            bucket: s3BucketBrowserApp.bucketName,
            policyDocument: new iam.PolicyDocument({
                statements: [
                    new iam.PolicyStatement({
                        effect: iam.Effect.ALLOW,
                        principals: [new iam.CanonicalUserPrincipal(cloudfrontOriginAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId)],
                        actions: ['s3:GetObject'],
                        resources: [
                            Fn.join('', [
                                s3BucketBrowserApp.bucketArn,
                                '/*',
                            ]),
                        ],
                    }),
                ],
            }),
        });

        utils.setNagSuppressRules(s3BucketAppPolicy,
            {
                id: 'AwsSolutions-S10',
                reason: 'HTTPS requirement not supported at this time.',
            }
        )

        // CloudFront Distribution
        /**
         * Distribution L2 construct is available for use
         * The solution, however, requires many cloudfront propeties not supported in L2 construct's native props
         * Thus, the L1 CfnCloudfront is used instead.
         */
        const cloudfrontDistribution = new cloudfront.CfnDistribution(this, 'MSAMAppBucketCloudFrontDistribution', {
            distributionConfig: {
                comment: Fn.join('', [
                    'CDN for ',
                    s3BucketBrowserApp.bucketName,
                ]),
                defaultCacheBehavior: {
                    targetOriginId: s3BucketBrowserApp.bucketName,
                    viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.HTTPS_ONLY,
                    minTtl: 3600,
                    defaultTtl: 7200,
                    maxTtl: 86400,
                    allowedMethods: [
                        'HEAD',
                        'GET',
                        'OPTIONS',
                    ],
                    cachedMethods: [
                        'HEAD',
                        'GET',
                        'OPTIONS',
                    ],
                    compress: true,
                    forwardedValues: {
                        queryString: false,
                        cookies: {
                            forward: 'none',
                        },
                        headers: [
                            'Origin',
                            'Access-Control-Request-Method',
                            'Access-Control-Request-Headers',
                        ],
                    },
                },
                defaultRootObject: 'index.html',
                enabled: true,
                origins: [{
                    domainName: s3BucketBrowserApp.bucketRegionalDomainName,
                    id: s3BucketBrowserApp.bucketName,
                    s3OriginConfig: {
                        originAccessIdentity: Fn.join('', [
                            'origin-access-identity/cloudfront/',
                            cloudfrontOriginAccessIdentity.originAccessIdentityId,
                        ]),
                    },
                }],
                priceClass: cloudfront.PriceClass.PRICE_CLASS_ALL,
                restrictions: {
                    geoRestriction: {
                        restrictionType: 'none',
                        locations: [],
                    },
                },
                viewerCertificate: {
                    cloudFrontDefaultCertificate: true,
                    minimumProtocolVersion: 'TLSv1.2_2019',
                },
                logging: {
                    bucket: s3BucketbrowserAppLogs.bucketRegionalDomainName,
                    prefix: 'cloudfront/',
                },
            }
        });
        this.MSAMBrowserURL = Fn.join('', [
            'https://',
            cloudfrontDistribution.attrDomainName,
        ]);
        utils.setNagSuppressRules(cloudfrontDistribution,
            {
                id: 'W70',
                id2: 'AwsSolutions-CFR4',
                reason: 'If the distribution uses the CloudFront domain name such as d111111abcdef8.cloudfront.net (you set CloudFrontDefaultCertificate to true), CloudFront automatically sets the security policy to TLSv1 regardless of the value that you set here.',
            },
            {
                id: 'AwsSolutions-CFR1',
                reason: 'Geo restriction not supported at this time.',
            },
            {
                id: 'AwsSolutions-CFR2',
                reason: 'AWS WAF integration not supported at this time.',
            }
        );

        // Web Content Lambda Function
        const webContentLambda = new lambda.Function(this, 'MSAMWebContentResource', {
            code: lambda.Code.fromBucket(
                s3.Bucket.fromBucketName(
                    this,
                    'ContentLambdaBucket',
                    Fn.join('-', ['%%BUCKET_NAME%%', Aws.REGION]),
                ),
                '%%SOLUTION_NAME%%/%%VERSION%%/CUSTOM_RESOURCE_FILE',
            ),
            environment: {
                BUILD_STAMP: 'DEV_0_0_0',
                ZIP_DIGEST: 'ZIP_DIGEST_VALUE',
                BUCKET_BASENAME: '%%BUCKET_NAME%%'
            },
            handler: 'cfn_bucket_loader.handler',
            description: 'MSAM Lambda for custom CloudFormation resource for loading web application',
            memorySize: 2560,
            role: props.WebIAMRole,
            runtime: lambda.Runtime.PYTHON_3_8,
            timeout: Duration.seconds(900),
        });
        utils.setNagSuppressRules(webContentLambda,
            {
                id: 'W58',
                reason: 'Role with AWSLambdaBasicExecutionRole defined in different template.'
            },
            {
                id: 'W89',
                reason: 'Lambda does not need to be in a VPC.'
            },
            {
                id: 'W92',
                reason: 'Lambda does not need ReservedConcurrentExecutions.'
            },
            {
                id: 'AwsSolutions-L1',
                reason: 'Latest runtime version not supported at this time.',
            },
        )

        // Web Content Custom Resource
        this.createCustomResource('MSAMWebContent', {
            serviceToken: webContentLambda.functionArn,
            properties: {
                BucketName: s3BucketBrowserApp.bucketName,
                StackName: Aws.STACK_NAME,
                BUILD_STAMP: 'DEV_0_0_0',
                ZIP_DIGEST: 'ZIP_DIGEST_VALUE',
            },
        });

        // Web Invalidation Resource Lambda Function
        const webInvalidationResourceLambda = new lambda.Function(this, 'MSAMWebInvalidationResource', {
            code: lambda.Code.fromBucket(
                s3.Bucket.fromBucketName(
                    this,
                    'InvalidationBucket',
                    Fn.join('-', ['%%BUCKET_NAME%%', Aws.REGION]),
                ),
                '%%SOLUTION_NAME%%/%%VERSION%%/CUSTOM_RESOURCE_FILE',
            ),
            environment: {
                BUILD_STAMP: 'DEV_0_0_0',
                ZIP_DIGEST: 'ZIP_DIGEST_VALUE',
                BUCKET_BASENAME: '%%BUCKET_NAME%%'
            },
            handler: 'cfn_invalidate_resource.handler',
            description: 'MSAM Lambda for custom resource for invalidating CloudFront after update',
            memorySize: 2560,
            role: props.WebIAMRole,
            runtime: lambda.Runtime.PYTHON_3_8,
            timeout: Duration.seconds(300),
        });
        utils.setNagSuppressRules(webInvalidationResourceLambda,
            {
                id: 'W58',
                reason: 'Role with AWSLambdaBasicExecutionRole defined in different template.'
            },
            {
                id: 'W89',
                reason: 'Lambda does not need to be in a VPC.'
            },
            {
                id: 'W92',
                reason: 'Lambda does not need ReservedConcurrentExecutions.'
            },
            {
                id: 'AwsSolutions-L1',
                reason: 'Latest runtime version not supported at this time.',
            },
        );

        // Web Invalidator Custom Resource
        this.createCustomResource('MSAMWebInvalidator', {
            serviceToken: webInvalidationResourceLambda.functionArn,
            properties: {
                BucketName: s3BucketBrowserApp.bucketName,
                StackName: Aws.STACK_NAME,
                BUILD_STAMP: 'DEV_0_0_0',
                ZIP_DIGEST: 'ZIP_DIGEST_VALUE',
                DistributionId: cloudfrontDistribution.attrId,
            },
        });
    }

    createCustomResource(id: string, props: CustomResourceProps) {
        return new CustomResource(this, id, props);
    }

    createS3BucketPolicy(id: string, props: s3.CfnBucketPolicyProps) {
        return new s3.CfnBucketPolicy(this, id, props);
    }
}
