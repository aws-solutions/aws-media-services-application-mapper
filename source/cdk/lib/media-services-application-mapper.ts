/*! Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
       SPDX-License-Identifier: Apache-2.0 */

import {
    Aws,
    Fn,
    Stack,
    StackProps,
    CfnParameter,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MsamDynamoDB } from './msam-dynamodb';
import { MsamIamRoles } from './msam-iam-roles';
import { MsamEvents } from './msam-events';
import { MsamBrowserApp } from './msam-browser-app'
import { MsamCore } from './msam-core';
import * as utils from './utils';

export class MediaServicesApplicationMapper extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(
            scope,
            id,
            {
                ...props,
                description: '(SO0048) Media Services Application Mapper Root Stack %%VERSION%%',
            }
        );

        /**
         * Parameters
         */
        const cacheItemTTL = new CfnParameter(this, 'CacheItemTTL', {
            default: '86400',
            description: 'This is the maximum time in seconds a cached item will remain if never updated (3600 = 1 hour, 86400 = 1 day, 604800 = 1 week).',
            type: 'String',
            allowedPattern: '\\S+',
            minLength: 1,
            constraintDescription: 'Please enter a value for this field.',
        });

        /**
         * Nested Stacks
         */

        // IAM Module Stack
        const iamModuleStack = new MsamIamRoles(this, 'IAMModuleStack');
        utils.addTemplateUrl(
            iamModuleStack,
            Fn.join('', [
                'https://%%BUCKET_NAME%%-',
                Aws.REGION,
                '.s3.',
                Aws.URL_SUFFIX,
                '/%%SOLUTION_NAME%%/%%VERSION%%/msam-iam-roles-release.template',
            ]),
        );

        // DynamoDB Module Stack
        const dynamoDBModuleStack = new MsamDynamoDB(this, 'DynamoDBModuleStack', {
            DynamoDBIAMRole: iamModuleStack.DynamoDBRole,
        });
        utils.addTemplateUrl(
            dynamoDBModuleStack,
            Fn.join('', [
                'https://%%BUCKET_NAME%%-',
                Aws.REGION,
                '.s3.',
                Aws.URL_SUFFIX,
                '/%%SOLUTION_NAME%%/%%VERSION%%/msam-dynamodb-release.template',
            ]),
        );

        // Core Module Stack
        const coreModuleStack = new MsamCore(this, 'CoreModuleStack', {
            CoreIAMRole: iamModuleStack.CoreRole,
            AlarmsTableName: dynamoDBModuleStack.AlarmsTable,
            ChannelsTableName: dynamoDBModuleStack.ChannelsTable,
            ContentTableName: dynamoDBModuleStack.ContentTable,
            EventsTableName: dynamoDBModuleStack.EventsTable,
            LayoutTableName: dynamoDBModuleStack.LayoutTable,
            SettingsTableName: dynamoDBModuleStack.SettingsTable,
            CloudWatchEventsTableName: dynamoDBModuleStack.CloudWatchEventsTable,
            NotesTableName: dynamoDBModuleStack.NotesTable,
            CacheItemTTL: cacheItemTTL.valueAsString,
            RootStackName: Aws.STACK_NAME,
        });
        utils.addTemplateUrl(
            coreModuleStack,
            Fn.join('', [
                'https://%%BUCKET_NAME%%-',
                Aws.REGION,
                '.s3.',
                Aws.URL_SUFFIX,
                '/%%SOLUTION_NAME%%/%%VERSION%%/msam-core-release.template',
            ]),
        );

        // Events Module Stack
        const eventsModuleStack = new MsamEvents(this, 'EventsModuleStack', {
            AlarmsTableName: dynamoDBModuleStack.AlarmsTable,
            EventsTableName: dynamoDBModuleStack.EventsTable,
            ContentTableName: dynamoDBModuleStack.ContentTable,
            CloudWatchEventsTableName: dynamoDBModuleStack.CloudWatchEventsTable,
            ItemTTL: cacheItemTTL.valueAsString,
            EventsTableRegion: Aws.REGION,
            EventsIAMRole: iamModuleStack.EventsRole,
        });
        utils.addTemplateUrl(
            eventsModuleStack,
            Fn.join('', [
                'https://%%BUCKET_NAME%%-',
                Aws.REGION,
                '.s3.',
                Aws.URL_SUFFIX,
                '/%%SOLUTION_NAME%%/%%VERSION%%/msam-events-release.template',
            ])
        );

        // Browser App Module Stack
        const browserAppModuleStack = new MsamBrowserApp(this, 'BrowserAppModuleStack', {
            WebIAMRole: iamModuleStack.WebRole,
        });
        utils.addTemplateUrl(
            browserAppModuleStack,
            Fn.join('', [
                'https://%%BUCKET_NAME%%-',
                Aws.REGION,
                '.s3.',
                Aws.URL_SUFFIX,
                '/%%SOLUTION_NAME%%/%%VERSION%%/msam-browser-app-release.template',
            ]),
        );

        // Service Catalog - App Registry
        utils.applyAppRegistry(this, [
            {
                stackName: 'DynamoDB',
                stack: dynamoDBModuleStack,
            },
            {
                stackName: 'IAM',
                stack: iamModuleStack,
            },
            {
                stackName: 'Events',
                stack: eventsModuleStack,
            },
            {
                stackName: 'Browser',
                stack: browserAppModuleStack,
            },
            {
                stackName: 'Core',
                stack: coreModuleStack,
            },
        ]);

        /**
         * Outputs
         */
        utils.createCfnOutput(this, 'MSAMBrowserURL', {
            description: 'URL for the MSAM browser application',
            value: browserAppModuleStack.MSAMBrowserURL,
        });

        utils.createCfnOutput(this, 'EndpointURL', {
            description: 'The endpoint needed by the MSAM browser application',
            value: coreModuleStack.EndpointURL,
        });

        utils.createCfnOutput(this, 'APIKeyID', {
            description: 'Link for retrieving API key needed by the MSAM browser application',
            value: coreModuleStack.APIKeyID,
        });
    }
}
