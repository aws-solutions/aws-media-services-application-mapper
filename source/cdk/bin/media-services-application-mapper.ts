#!/usr/bin/env node
/*! Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
       SPDX-License-Identifier: Apache-2.0 */

import * as cdk from 'aws-cdk-lib';
import { MediaServicesApplicationMapper } from '../lib/media-services-application-mapper';
import { AwsSolutionsChecks } from 'cdk-nag';

const cdkApplication = new cdk.App();

(function createMediaServicesApplicationMapperStack(app) {
    return new MediaServicesApplicationMapper(app, 'MediaServicesApplicationMapper');
})(cdkApplication);

//cdk nag
cdk.Aspects.of(cdkApplication).add(new AwsSolutionsChecks({ verbose: true }));
