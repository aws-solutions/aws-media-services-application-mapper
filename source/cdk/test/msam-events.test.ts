/*! Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
       SPDX-License-Identifier: Apache-2.0 */

import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MediaServicesApplicationMapper } from '../lib/media-services-application-mapper';

test('msam-events snapshot test', () => {
    const app = new cdk.App();
    const stack = new MediaServicesApplicationMapper(app, 'MediaServicesApplicationMapper');
    const template = Template.fromStack(stack.node.findChild('EventsModuleStack') as cdk.NestedStack);
    
    expect(template).toMatchSnapshot();
});
