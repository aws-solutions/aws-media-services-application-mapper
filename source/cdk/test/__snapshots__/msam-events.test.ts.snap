// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`msam-events snapshot test 1`] = `
{
  "Description": "Media Services Application Mapper Event Capture %%VERSION%%",
  "Mappings": {
    "SolutionId": {
      "UserAgent": {
        "Extra": "AwsSolution/SO0048/%%VERSION%%",
      },
    },
  },
  "Parameters": {
    "referencetoMediaServicesApplicationMapperCacheItemTTL6DD8B4F7Ref": {
      "Type": "String",
    },
    "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackAlarmsF6F9E932Ref": {
      "Type": "String",
    },
    "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackCloudWatchEvents6ACDEB1ARef": {
      "Type": "String",
    },
    "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackContent622D5AD1Ref": {
      "Type": "String",
    },
    "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackEventsFE64F5A1Ref": {
      "Type": "String",
    },
    "referencetoMediaServicesApplicationMapperIAMModuleStackNestedStackIAMModuleStackNestedStackResourceF97053D6OutputsMediaServicesApplicationMapperIAMModuleStackEventsRole2AFF40D1Arn": {
      "Type": "String",
    },
  },
  "Resources": {
    "AlarmUpdater6FDC3F1D": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-L1",
              "reason": "Latest runtime version not supported at this time.",
            },
          ],
        },
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "Role with CloudWatch Logs permissions defined in different template.",
            },
            {
              "id": "W92",
              "reason": "Lambda does not need ReservedConcurrentExecutions.",
            },
            {
              "id": "W89",
              "reason": "Lambda does not need to be in a VPC.",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Join": [
              "-",
              [
                "%%BUCKET_NAME%%",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "%%SOLUTION_NAME%%/%%VERSION%%/events_DEV_0_0_0.zip",
        },
        "Description": "MSAM Lambda for handling CloudWatch alarm state change events.",
        "Environment": {
          "Variables": {
            "ALARMS_TABLE_NAME": {
              "Ref": "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackAlarmsF6F9E932Ref",
            },
            "BUILD_STAMP": "DEV_0_0_0",
            "EVENTS_TABLE_REGION": {
              "Ref": "AWS::Region",
            },
            "SOLUTION_ID": {
              "Fn::FindInMap": [
                "SolutionId",
                "UserAgent",
                "Extra",
              ],
            },
          },
        },
        "Handler": "cloudwatch_alarm.lambda_handler",
        "MemorySize": 2560,
        "Role": {
          "Ref": "referencetoMediaServicesApplicationMapperIAMModuleStackNestedStackIAMModuleStackNestedStackResourceF97053D6OutputsMediaServicesApplicationMapperIAMModuleStackEventsRole2AFF40D1Arn",
        },
        "Runtime": "python3.8",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlarmUpdaterAlarmChangeStateEventsA22ABD95": {
      "Properties": {
        "EventPattern": {
          "detail-type": [
            "CloudWatch Alarm State Change",
          ],
          "source": [
            "aws.cloudwatch",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "AlarmUpdater6FDC3F1D",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "AlarmUpdaterAlarmChangeStateEventsAllowEventRuleMediaServicesApplicationMapperEventsModuleStackAlarmUpdater706835E134F56820": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AlarmUpdater6FDC3F1D",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "AlarmUpdaterAlarmChangeStateEventsA22ABD95",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "Collector9EBA7CF5": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-L1",
              "reason": "Latest runtime version not supported at this time.",
            },
          ],
        },
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "Role with CloudWatch Logs permissions defined in different template.",
            },
            {
              "id": "W92",
              "reason": "Lambda does not need ReservedConcurrentExecutions.",
            },
            {
              "id": "W89",
              "reason": "Lambda does not need to be in a VPC.",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Join": [
              "-",
              [
                "%%BUCKET_NAME%%",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": "%%SOLUTION_NAME%%/%%VERSION%%/dynamodb_resource_DEV_0_0_0.zip",
        },
        "Description": "MSAM Lambda for handling CloudWatch event notifications",
        "Environment": {
          "Variables": {
            "BUILD_STAMP": "DEV_0_0_0",
            "CLOUDWATCH_EVENTS_TABLE_NAME": {
              "Ref": "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackCloudWatchEvents6ACDEB1ARef",
            },
            "CONTENT_TABLE_NAME": {
              "Ref": "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackContent622D5AD1Ref",
            },
            "EVENTS_TABLE_NAME": {
              "Ref": "referencetoMediaServicesApplicationMapperDynamoDBModuleStackNestedStackDynamoDBModuleStackNestedStackResource588058D0OutputsMediaServicesApplicationMapperDynamoDBModuleStackEventsFE64F5A1Ref",
            },
            "EVENTS_TABLE_REGION": {
              "Ref": "AWS::Region",
            },
            "ITEM_TTL": {
              "Ref": "referencetoMediaServicesApplicationMapperCacheItemTTL6DD8B4F7Ref",
            },
            "SOLUTION_ID": {
              "Fn::FindInMap": [
                "SolutionId",
                "UserAgent",
                "Extra",
              ],
            },
          },
        },
        "Handler": "media_events.lambda_handler",
        "MemorySize": 2560,
        "Role": {
          "Ref": "referencetoMediaServicesApplicationMapperIAMModuleStackNestedStackIAMModuleStackNestedStackResourceF97053D6OutputsMediaServicesApplicationMapperIAMModuleStackEventsRole2AFF40D1Arn",
        },
        "Runtime": "python3.8",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MediaEvents346D729A": {
      "Properties": {
        "EventPattern": {
          "source": [
            "aws.medialive",
            "aws.mediapackage",
            "aws.mediastore",
            "aws.mediatailor",
            "aws.mediaconnect",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "Collector9EBA7CF5",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MediaEventsAllowEventRuleMediaServicesApplicationMapperEventsModuleStackCollector0814B3F9E4AA863F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Collector9EBA7CF5",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "MediaEvents346D729A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
